### ISSUE TITLE
queryWithParams can not use all connection in the pool

### ISSUE THREAD

**Issue by VoxT** [vert-x3/vertx-jdbc-client#164]
### Version

* vert.x core: `3.7.1`
* vert.x jdbc-client: `3.7.1`
* jdbc driver: `6.0.6`
* RDBMS: `mysql server`

### General info

* What is your jdbc driver: `mysql-connector-java`
* What is your RDBMS server: `mysql`
* What is your connection string (no user/passwords please!): j`dbc:mysql://127.0.0.1:3306/db _dev?autoReconnect=true&useLegacyDatetimeCode=false&serverTimezone=UTC`

### Context

I execute a query to the db, and there is two different way that i use jdbc-client and produce two performance result that should be the same.

- create pool
```java
HikariConfig config = new HikariConfig();
    config.setDriverClassName(mySQLConfig.getDriver());
    config.setJdbcUrl(mySQLConfig.getUrl());
    config.setUsername(mySQLConfig.getUser());
    config.setPassword(mySQLConfig.getPassword());
    config.setMaximumPoolSize(mySQLConfig.getPoolSize());
    config.setAutoCommit(isAutoCommit);
    config.addDataSourceProperty("useServerPrepStmts", "" + mySQLConfig.isUseServerPrepStmts());
    config.addDataSourceProperty("cachePrepStmts", "" + mySQLConfig.isCachePrepStmts());
    config.addDataSourceProperty("prepStmtCacheSize", "" + mySQLConfig.getPrepStmtCacheSize());
    config.addDataSourceProperty(
        "prepStmtCacheSqlLimit", "" + mySQLConfig.getPrepStmtCacheSqlLimit());
    config.addDataSourceProperty("maxLifetime", "" + mySQLConfig.getMaxLifetimeMillis());

    try {
      logger.info("Connecting to datasource={}", JsonProtoUtils.printGson(mySQLConfig));
      DataSource dataSource = new HikariDataSource(config);

      mySQLClient = JDBCClient.create(vertx, dataSource);
      if (mySQLClient == null) {
        throw new Exception("mySQLClient == null");
      }
    } catch (Exception e) {
      future.fail(e);
      return future;
    }
```

- method 1: query using getConnection
```java
mySQLClient.getConnection(
          resConnection -> {
            if (resConnection.failed()) {
              result.fail(resConnection.cause());
              return;
            }

            SQLConnection connection = resConnection.result();
            connection.queryWithParams(
                prepareStm,
                params,
                selectRes -> {
                  if (selectRes.failed()) {
                    accountFuture.fail(selectRes.cause());
                  } else {
                    getResult(selectRes.result());
                  }
                  connection.close();
                });
          });
```

- method 2: query using SQLClient.queryWithParams
```java
mySQLClient
        .queryWithParams(prepareStm, params, asyncResultHandler);
```
- and there is result of prometheus metric when i run a benchmark, with connection pool config is 20:
  * method 1: vertx_pool_inUse{pool_type="datasource"}: 20
  * method 2: vertx_pool_inUse{pool_type="datasource"}: 1

and of course, method 1 has req/s much higher than method 2.

I see that both methods above use the  same method
```java
// class JDBCClientImpl
public SQLClient getConnection(Handler<AsyncResult<SQLConnection>> handler) {
    Context ctx = this.vertx.getOrCreateContext();
    this.getConnection(ctx, (ar) -> {
      ctx.runOnContext((v) -> {
        handler.handle(ar);
      });
    });
    return this;
  }
```

then why it makes a huge difference?

**Comment by pendula95** [comment]
Bump here, I see same behaviour.
My pool in use is always 1 when using mySQLClient.queryWithParams and if I use SQLConnection.queryWithParams I get full pool used and better performance.
Vertx version: 3.9.2
Can this be confirmed?

**Comment by vietj** [comment]
can you investigate @pmlopes ?

**Comment by pendula95** [comment]
Here is a small reproducer:

```
    public static void main(String[] args) {

        Vertx vertx = Vertx.vertx(new VertxOptions().setMetricsOptions(
                new MicrometerMetricsOptions()
                        .addLabels(Label.POOL_NAME)
                        .setPrometheusOptions(new VertxPrometheusOptions().setEnabled(true))
                        .setEnabled(true)));

        Router router = Router.router(vertx);
        router.route("/metrics").handler(PrometheusScrapingHandler.create());
        vertx.createHttpServer().requestHandler(router).listen(8888);

        JsonObject dbSettings = new JsonObject()
                .put("url", "jdbc:postgresql://127.0.0.1:5434/postgres?currentSchema=public")
                .put("user", "postgres")
                .put("password", "postgres")
                .put("max_pool_size", 30);

        JDBCClient client1 = JDBCClient.createShared(vertx, dbSettings, "Client1");
        JDBCClient client2 = JDBCClient.createShared(vertx, dbSettings, "Client2");

        vertx.setPeriodic(5, handler -> {
            client1.query("SOME QUERY'", result -> {
                if (result.failed()) {
                    result.cause().printStackTrace();
                }
                try {
                    Thread.sleep(20);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            });
        });

        vertx.setPeriodic(5, handler -> {
            client2.getConnection(connection -> {
                if (connection.succeeded()) {
                    connection.result().query("SOME QUERY'", result -> {
                        if (result.failed()) {
                            result.cause().printStackTrace();
                        }
                    });
                } else {
                    connection.cause().printStackTrace();
                }
                try {
                    Thread.sleep(20);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                connection.result().close();
            });

        });

        vertx.setPeriodic(5000, handler -> {
            MetricsService metricsService = MetricsService.create(vertx);
            System.out.println(metricsService.getMetricsSnapshot("vertx.pool.inUse").encodePrettily());
        });
    }
```

For some reason for client1 value of vertx.pool.inUse is always 0
For client 2 the values changes constantly.
Maybe this has something to do with metrics and not pools not used.

**Comment by pendula95** [comment]
I have changed the code so I can put more load on it and now the difference is obvious, when you not get connection explicitly you are not using the pool.
```
router.get("/bench1").handler(routingContext -> {
            client1.query("SELECT * FROM operator WHERE code = 'HK_IPIFICATION'", result -> {
                if (result.failed()) {
                    result.cause().printStackTrace();
                    routingContext.response().setStatusCode(500).end();
                } else {
                    client1Counter.incrementAndGet();
                    routingContext.response().setStatusCode(200).end();
                }
            });
        });

        router.get("/bench2").handler(routingContext -> {
            client2.getConnection(connection -> {
                if (connection.succeeded()) {
                    connection.result().query("SELECT * FROM operator WHERE code = 'HK_IPIFICATION'", result -> {
                        if (result.failed()) {
                            result.cause().printStackTrace();
                            routingContext.response().setStatusCode(500).end();
                        } else {
                            client2Counter.incrementAndGet();
                            routingContext.response().setStatusCode(200).end();
                        }
                    });
                } else {
                    connection.cause().printStackTrace();
                    routingContext.response().setStatusCode(500).end();
                }
                connection.result().close();
            });
        });
``` 
After running benchmarking this with bombardier (each method, 3 tries for 30s) I get following:
Bench1: 
used pool size 1
Total requests served 52558
Bench2:
used pool size 20-30
Total requests server 1064291

So this is an issue with JDBC client and not bad metric count.

**Comment by jinwik** [comment]
when using jdbclient.queryXX，requests run in "vertx-jdbc-service-get-connection-thread" thread which is a single thread!
when using getConnection and con.queryXX, requets run in worker-threads(executeBlocking).

**Comment by pendula95** [comment]
This is because JDBCConnectionImpl will call method execute from AbstractJDBCAction which wraps it around executeBlocking [here](https://github.com/vert-x3/vertx-jdbc-client/blob/670670dbb48f4fe6159a49843ba4b8fcb1c60d92/src/main/java/io/vertx/ext/jdbc/impl/actions/AbstractJDBCAction.java#L66)
On the other had JDBCClientImpl will execute it from extensions of AbstractJDBCAction which is not wrapped around executeBlocking. Example for query call [here](https://github.com/vert-x3/vertx-jdbc-client/blob/670670dbb48f4fe6159a49843ba4b8fcb1c60d92/src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCQuery.java#L42 )
I think I can make a PR for this, dnk if @pmlopes already started

**Comment by pmlopes** [comment]
@pendula95 please do!

**Comment by pendula95** [comment]
I have checkout 3.92 tag and applied changes to that head but looks like there are conflicts merging to master.

**Comment by tsegismont** [comment]
Closed in #213
